# Машинное обучение, ФКН ВШЭ

## Практическое задание 1

### Общая информация
Дата выдачи: 06.09.2019  

Мягкий дедлайн: 23:59MSK 15.09.2019 (за каждый день просрочки снимается 1 балл)

Жесткий дедлайн: 23:59MSK 17.09.2019  

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В каждом разделе вам предлагается выполнить несколько заданий. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотекам pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Задания сдаются через систему Anytask. Инвайт можно найти на странице курса. Присылать необходимо ноутбук с выполненным заданием. 

Для удобства проверки самостоятельно посчитайте свою максимальную оценку (исходя из набора решенных задач) и укажите ниже.

Оценка: xx.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написаем production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [80]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

А это ___ячейка с текстом___.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [81]:
# your code
19**23

257829627945307727248226067259

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. #Делать 
##заголовки 
###разного уровня
3. Выделять *текст* <s>при</s> **необходимости**
4. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки:
<img src="https://st2.depositphotos.com/1177973/9266/i/950/depositphotos_92668716-stock-photo-red-cat-with-computer-keyboard.jpg" style="width: 400px">

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных. С этой библиотекой мы сейчас и познакомимся.

Основной объект в pandas это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [195]:
%pylab inline  
# import almost all we need
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Данные можно скачать [отсюда](https://www.dropbox.com/s/5qq94wzmbw4e54r/data.csv?dl=0).

#### 1. [0.5 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [196]:
data = pd.read_csv("data.csv")
print('Table rows size = %s , '
      'columns size = %s' % 
      (data.shape[0], data.shape[1]))
data.tail(10)

Table rows size = 4622 , columns size = 5


order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25   
4613                                                NaN     $2.15   
4614                                                NaN     $1.50   
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

Данная таблица показывает список заказов. Каждый заказ имеет свой уникальный id. У каждого заказа есть выбранные товары, их колличество, цена и описание.

#### 2. [0.25 баллов] Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [197]:
print('Orders count = %s' % data['order_id'].max())
print('Unique goods = %s' % len(data['item_name'].unique()))

Orders count = 1834
Unique goods = 50


#### 3. [0.25 баллов] Есть ли в данных пропуски? В каких колонках? 

In [198]:
for column in data:
    print('Empty fields count in %s is %s' % (column, data[column].isnull().sum()))

Empty fields count in order_id is 0
Empty fields count in quantity is 0
Empty fields count in item_name is 0
Empty fields count in choice_description is 1246
Empty fields count in item_price is 0


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [199]:
default_values = {'order_id': 0, 'quantity': 0, 'item_name': '',
                  'choice_description': '', 'item_price': ''}
data = data.fillna(value=default_values)
data.tail(10)

order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25   
4613                                                        $2.15   
4614                                                        $1.50   
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                        $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

#### 4. [0.5 баллов] Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [200]:
data['item_price_float'] = data['item_price'].str.replace('$', '').astype(float)
data.tail(10)

order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  \
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25    
4613                                                        $2.15    
4614                                                        $1.50    
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75    
4616                                                        $4.45    
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75    
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75    
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25    
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75    
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75    

      item_price_float  
4612              9.25  
4613              2.15  
4614              1.50  
4615              8.75  
4616              4.45  
4617             11.75  
4618             11.75  
4619             11.25  
4620              8.75  
4621              8.75

Какая средняя/минимальная/максимальная цена у товара? 

In [201]:
print('Min price = %s' % data['item_price_float'].min())
print('Max price = %s' % data['item_price_float'].max())
print('Average price = %s' % data['item_price_float'].mean())

Min price = 1.09
Max price = 44.25
Average price = 7.464335785374397


Удалите старую колонку с ценой.

In [202]:
data = data.drop(['item_price'], axis=1)
data.tail(10)

order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description  item_price_float  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...              9.25  
4613                                                                 2.15  
4614                                                                 1.50  
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]              8.75  
4616                                                                 4.45  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...             11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...             11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...             11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...              8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...              8.75

#### 5. [0.25 баллов] Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [203]:
def calc_one_item_cost(row):
    return float(row['item_price_float'] / row['quantity'])
data['one_item_cost'] = data.apply(calc_one_item_cost, axis=1, result_type='expand')
data.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price_float  \
0                                                                 2.39   
1                                       [Clementine]              3.39   
2                                            [Apple]              3.39   
3                                                                 2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...             16.98   

   one_item_cost  
0           2.39  
1           3.39  
2           3.39  
3           2.39  
4           8.49

In [205]:
data.sort_values('one_item_cost').drop_duplicates('item_name', 'last').nlargest(5, 'one_item_cost')

order_id  quantity            item_name  \
4239      1692         1  Carnitas Salad Bowl   
2442       970         1  Barbacoa Salad Bowl   
2401       953         1     Steak Salad Bowl   
3694      1480         1   Steak Crispy Tacos   
2033       818         1        Carnitas Bowl   

                                     choice_description  item_price_float  \
4239  [Tomatillo Green Chili Salsa, [Black Beans, Ch...             11.89   
2442  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...             11.89   
2401  [Fresh Tomato Salsa, [Fajita Vegetables, Guaca...             11.89   
3694  [Tomatillo Green Chili Salsa, [Cheese, Sour Cr...             11.75   
2033  [Tomatillo Red Chili Salsa, [Fajita Vegetables...             11.75   

      one_item_cost  
4239          11.89  
2442          11.89  
2401          11.89  
3694          11.75  
2033          11.75

In [206]:
data.sort_values('one_item_cost').drop_duplicates('item_name', 'first').nsmallest(5, 'one_item_cost')

order_id  quantity          item_name choice_description  \
3364      1349         2        Canned Soda        [Coca Cola]   
818        338         3      Bottled Water                      
4200      1676         1  Canned Soft Drink             [Coke]   
4057      1624         1      Side of Chips                      
4509      1793         1              Chips                      

      item_price_float  one_item_cost  
3364              2.18           1.09  
818               3.27           1.09  
4200              1.25           1.25  
4057              1.69           1.69  
4509              1.99           1.99

#### 6. [0.5 баллов] Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [215]:
def calc(row):
    if row['item_name']  == 'Chicken Bowl' and row['quantity'] > 1:
        return True
print('More than one Chicken Bowl orders count = %s' % data.apply(calc, axis=1, result_type='expand').sum())

More than one Chicken Bowl orders count = 33


#### 7. [0.5 баллов] Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [216]:
print('Average price = %s' % data.groupby(['order_id'])['item_price_float'].sum().mean())
print('Average items count = %s' % data.groupby(['order_id'])['quantity'].sum().mean())

Average price = 18.81142857142869
Average items count = 2.711014176663032


#### 8. [0.25 баллов] Сколько заказов содержали ровно 1 товар?

In [228]:
tmp_data = data.groupby(['order_id'])['quantity'].sum()
print('Only one good orders count = %s' % tmp_data.where(tmp_data == 1).count())

Only one good orders count = 56


#### 9. [0.25 баллов] Какая самая популярная категория товара? 

In [ ]:
# your code

#### 10. [0.5 баллов] Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой? 

In [ ]:
# your code

#### 11. [0.75 баллов] В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно среднюю стоимость заказа, добавить ее в исходные данные и сделать необходимые проверки.

*Данный комментарий стоит воспринимать как подсказку к одному из вариантов решений задания. Если в вашем варианте решения он не нужнен, это не страшно*

In [ ]:
# your code

#### 12. [0.75 баллов] Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [ ]:
# your code

#### 13. [0.75 баллов] Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [ ]:
data = np.random.rand(10, 3)

# your code

Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [ ]:
# your code

## 2. Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать.

In [ ]:
%matplotlib inline  # нужно для отображения графиков внутри ноутбука
import matplotlib.pyplot as plt

На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [ ]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])

#### 1. [1 балл] Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name. 

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [ ]:
# your code

#### 2. [1 балл] Постройте график зависимости суммы покупок от дней.

In [ ]:
# your code

#### 3. [1 балл] Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
# your code

#### 4. [1 балл] Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 